File: Ruby/sorting.rb
module Sorting
	def self.bubblesort(arr)
		c = true
		while c
			c = false
			for j in 0...(arr.length-1)
				if arr[j] > arr[j+1]
					arr[j],arr[j+1] = arr[j+1],arr[j]
					c = true
				end # end if
			end # end for
		end # end for
	end

	def self.insertionsort(arr)
		arr.length.times do |x|

			for i in 1...arr.length

				if arr[i-1] > arr[i]
					hold = arr[i]
					pos = i

					while pos > 0 && arr[pos - 1] > hold
						arr[pos] = arr[pos - 1]
						pos -= 1
					end # end while
					arr[pos] = hold
				end # end if
			end # end for
		end	# end .times
	end # end def

	def self.selectionsort(arr)
		for i in 0...arr.length

			min_pos = i
			for j in i+1...arr.length
				min_pos = j if arr[j] < arr[min_pos]
			end # end for

			arr[min_pos], arr[i] = arr[i], arr[min_pos]
		end # end for
	end

	def self.mergesort(arr)
		return arr if arr.length <= 1
		if arr.length == 2
			arr[0],arr[1] = arr[1],arr[0] if arr[0] > arr[1]
			return arr
		end

		left = mergesort(arr[0...arr.length/2].dup)
		right = mergesort(arr[arr.length/2...arr.length].dup)

		i = 0
		until left.empty? || right.empty?
			arr[i] = (left.first < right.first ? left.shift : right.shift)
			i+=1
		end

		if !left.empty?
			i.upto(arr.length-1){|j| arr[j] = left.shift }
		elsif !right.empty?
			i.upto(arr.length-1){|j| arr[j] = right.shift }
		end

		return arr
	end

	def self._quicksort(arr, _begin, _end)
		return if _end - _begin <= 1

		a = _begin
		b = _end - 1

		pivot = arr[ b ]

		while (b > a)
			if arr[b-1] > pivot
				arr[b] = arr[b-1]
				b-=1
			else
				arr[b-1],arr[a] = arr[a],arr[b-1]
				a+=1
			end
		end

		arr[b] = pivot

		self._quicksort(arr, _begin, b)
		self._quicksort(arr, b+1, _end)
	end

	def self.quicksort(arr)
		self._quicksort(arr, 0, arr.length)
	end
end

File: Java/Sorting.java
import java.util.ArrayList;

public class Sorting<T extends Comparable<T>> {
	public static <T extends Comparable<T>> void
	bubblesort(ArrayList<T> arr) {
		boolean c = true;
		while ( c && !(c = false))
			for (int j = 0 ; j < arr.size()-1 ; j++)
				if ( arr.get(j).compareTo( arr.get(j+1) ) > 0 && (c = true))
					arr.set(j, swap( arr.get(j+1), arr.set(j+1, arr.get(j)) ));
	}

	public static <T extends Comparable<T>> void
	insertionsort(ArrayList<T> arr) {
		for (int j = 1 ; j < arr.size() ; j++) {
			if (arr.get(j).compareTo(arr.get(j-1)) < 0) {
				T hold = arr.get(j);
				int k = j;
				while(k > 0 && arr.get(k-1).compareTo(hold) > 0)
					arr.set(k, arr.get(--k));

				arr.set(k, hold);
			} /* end if */
		} /* end for */
	}

	public static <T extends Comparable<T>> void
	selectionsort(ArrayList<T> arr) {
		for (int i = 0 ; i < arr.size() ; i++) {
			int min_pos = i;
			for (int j = i+1 ; j < arr.size() ; j++)
				min_pos = arr.get(j).compareTo(arr.get(min_pos)) < 0 ? j : min_pos;

			if (min_pos != i)
				arr.set(i, swap( arr.get(min_pos), arr.set(min_pos, arr.get(i)) ));

		} /* end for */
	}

	public static <T extends Comparable<T>> void
	mergesort(ArrayList<T> arr) {
		if (arr.size() == 2) {
			if (arr.get(0).compareTo(arr.get(1)) > 0)
				arr.set(0, swap( arr.get(1), arr.set(1, arr.get(0)) ));
		}
		if (arr.size() <= 2) return;

		ArrayList<T> left,right;
		left = new ArrayList<T>();
		right = new ArrayList<T>();

		for (int i = 0; i < arr.size()/2; i++)
			left.add(arr.get(i));
		for (int i = arr.size()/2; i < arr.size(); i++)
			right.add(arr.get(i));

		mergesort(left); mergesort(right);

		arr.clear();
		while (!left.isEmpty() && !right.isEmpty())
			if (left.get(0).compareTo(right.get(0)) < 0)
				arr.add(left.remove(0));
			else
				arr.add(right.remove(0));

		if (!left.isEmpty())
			arr.addAll(left);
		else if (!right.isEmpty())
			arr.addAll(right);

	}

	public static <T extends Comparable<T>> void
	quicksort(ArrayList<T> arr) {
		quicksort(arr, 0 , arr.size());
	}

	private static <T extends Comparable<T>> void
	quicksort(ArrayList<T> arr, int begin, int end) {
		if (end <= begin) return;

		T pivot = arr.get(end-1);
		int a = begin, b = end-1;

		while(b > a) {
			if ( arr.get(b-1).compareTo(pivot) > 0 ) {
				arr.set(b , arr.get(--b));
			} else {
				arr.set(a , swap(arr.get(b-1) , arr.set(b-1 , arr.get(a) ) ));
			}
		}

		arr.set(b, pivot);
		quicksort(arr , begin , b);
		quicksort(arr , b + 1 , end);
	}

	public static <T> T
	swap(T a, T b) {
		/*
			https://stackoverflow.com/questions/2393906/how-do-i-make-my-swap-function-in-java/20600020#20600020
			AND THEY SAID IT COULDN'T BE DONE.
		*/
		return a;
	}
}
